library/x509_create.c

```diff
@@ -86,73 +97,345 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
     return( cur->oid );
 }
 
+static int x509_dn_check_special_chars( const char **cur, const char *end )
+{
+    const char *c = *cur;
+    int ret = 0;
+
+    /* Check for valid escaped characters */
+    if( c == end || *c != ',' )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+
+    return( ret );
+}
+
 int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
 {
     int ret = 0;
-    const char *s = name, *c = s;
-    const char *end = s + strlen( s );
+    /*
+     * `start` is a pointer into the 'name' buffer.
+     * Depending on the parsing state, it either points to the beginning of
+     * the currently parsed DN OID tag or the currently parsed DN value.
+     * `cur` points to the current character in the string being parsed.
+     * `d` acts as a write pointer into the temporary 'data' buffer which holds
+     * the unescaped and unquoted DN values found in 'name'.
+     */
+    const char *start = name , *cur = start;
+    const char *end = start + strlen( start );
     const char *oid = NULL;
-    int in_tag = 1;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
+    x509_dn_parse_state dn_cur_state = dn_before_tag;
+    size_t value_len = 0;
 
     /* Clear existing chain if present */
     mbedtls_asn1_free_named_data_list( head );
 
-    while( c <= end )
+    while( cur <= end )
     {
-        if( in_tag && *c == '=' )
+        switch( dn_cur_state )
         {
-            if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+            /*
+             * Currently starting a new DN tag
+             */
+            case dn_before_tag:
             {
-                ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
-                goto exit;
+                /*
+                 * Skip whitespaces
+                 */
+                while( cur < end && *cur == ' ' )
+                {
+                    cur++;
+                    start++;
+                }
+
+                if( *cur == '=' )
+                {
+                    ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
+                    goto exit;
+                }
+                else
+                {
+                    dn_cur_state = dn_in_tag;
+                    continue;
+                }
             }
-
-            s = c + 1;
-            in_tag = 0;
-            d = data;
-        }
-
-        if( !in_tag && *c == '\\' && c != end )
-        {
-            c++;
-
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+            break;
```

Comment:
If you mean literraly `A=B`, then `x509_attr_descr_from_name()` should fail, befcause `A` is not a know OID.

If you mean because no spaces, then this is valid